lib/maths/CDataFrameUtils.cc

```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+                                    [&columnIsCategorical](std::size_t i) {
+                                        return columnIsCategorical[i] == false;
+                                    }),
+                     columnMask.end());
+
+    if (frame.numberRows() == 0 || columnMask.empty()) {
+        return result;
+    }
+
+    auto results = frame.readRows(
+        numberThreads,
+        core::bindRetrievableState(
+            [&](TDoubleVecVec& counts, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    for (std::size_t i : columnMask) {
+                        std::size_t id{static_cast<std::size_t>((*row)[i])};
+                        counts[i].resize(std::max(counts[i].size(), id + 1), 0.0);
+                        counts[i][id] += 1.0;
+                    }
+                }
+            },
+            TDoubleVecVec(frame.numberColumns())));
+
+    if (results.second == false) {
+        HANDLE_FATAL(<< "Internal error: failed to calculate category"
+                     << " frequencies. Please report this problem.");
+        return result;
+    }
+
+    for (const auto& counts_ : results.first) {
+        for (std::size_t i = 0; i < counts_.s_FunctionState.size(); ++i) {
+            result[i].resize(counts_.s_FunctionState[i].size());
+        }
+    }
+    for (const auto& counts_ : results.first) {
+        for (std::size_t i = 0; i < counts_.s_FunctionState.size(); ++i) {
+            for (std::size_t j = 0; j < counts_.s_FunctionState[i].size(); ++j) {
+                result[i][j] += counts_.s_FunctionState[i][j] /
+                                static_cast<double>(frame.numberRows());
+            }
+        }
+    }
+
+    return result;
+}
+
+CDataFrameUtils::TDoubleVecVec
+CDataFrameUtils::meanValueOfTargetForCategories(std::size_t numberThreads,
+                                                const core::CDataFrame& frame,
+                                                TSizeVec columnMask,
+                                                std::size_t targetColumn) {
+
+    TDoubleVecVec result(frame.numberColumns());
+
+    if (targetColumn >= frame.numberColumns()) {
+        HANDLE_FATAL(<< "Internal error: target column out of bounds '"
+                     << targetColumn << " >= " << frame.numberColumns()
+                     << "'. Please report this problem.");
+        return result;
+    }
+
+    const auto& columnIsCategorical = frame.columnIsCategorical();
+    columnMask.erase(std::remove_if(columnMask.begin(), columnMask.end(),
+                                    [&columnIsCategorical](std::size_t i) {
+                                        return columnIsCategorical[i] == false;
+                                    }),
+                     columnMask.end());
+
+    if (frame.numberRows() == 0 || columnMask.empty()) {
+        return result;
+    }
```

Comment:
I think I prefer to have these functions take care of this (rather than every calling site), but definitely agree factoring out is a good idea. This was done in [25fa629](https://github.com/elastic/ml-cpp/pull/539/commits/25fa6295e9f6dc23993cc290699f6409f5f4e663).